src/PostgREST/Main.hs

```diff
@@ -67,17 +80,45 @@ main = do
     H.poolSettings (fromIntegral $ configPool conf) 30
   pool :: H.Pool P.Postgres <- H.acquirePool pgSettings poolSettings
 
-  resOrError <- H.session pool isServerVersionSupported
+  supportedOrError <- H.session pool isServerVersionSupported
   either (fail . show)
     (\supported ->
       unless supported $
         fail "Cannot run in this PostgreSQL version, PostgREST needs at least 9.2.0"
-    ) resOrError
+    ) supportedOrError
+
+    -- read the structure of the database
+    -- read the structure of the database
+  let txParam = Just (H.ReadCommitted, Just True)
+
+  tblsRes <-  H.session pool $ H.tx txParam alltables
+  let allTables = either (fail . show) id tblsRes
+
+  relsRes <-  H.session pool $ H.tx txParam allrelations
+  let allRelations = either (fail . show) id relsRes
+
+  colsRes <-  H.session pool $ H.tx txParam $ allcolumns allRelations
+  let allColumns = either (fail . show) id colsRes
+
+  pkRes <-  H.session pool $ H.tx txParam allprimaryKeys
```

Comment:
How about getting tblsRes, relsRes, colsRes, and pkRes from within the same session and transaction? I think it would be lower overhead. How about this:

```haskell
metadata <- H.session pool $ H.tx txParam $ do
  tabs <- alltables
  rels <- allrelations
  cols <- allcolumns rels
  keys <- allprimaryKeys
  return (tabs, rels, cols, keys)
                                               
dbstructure <- case metadata of
  Left e -> fail $ show e
  Right (tabs, rels, cols, keys) ->
    return $ DbStructure {
        tables=tabs
      , columns=cols
      , relations=rels
      , primaryKeys=keys
      }
```

Also the all-lowercase names from PgStructure (like `alltables`) don't match the camelCasing of the rest of the code.